src/providers/structure.ts

```diff
@@ -68,6 +56,419 @@ export class SectionNodeProvider implements vscode.TreeDataProvider<Section> {
+     * @param commands The LaTeX commands to be extracted.
+     * @param depths The correspondance of section types and depths. Start from
+     * zero is the top-most section (e.g., chapter). -1 is reserved for
+     * non-section commands.
+     * @returns A flat array of {@link Section} of this file.
+     */
+    private async buildLaTeXSectionFromFile(file: string, subFile: boolean, filesBuilt: Set<string>, commands: {cmds: string[], envs: string[], secs: string[]}, depths: {[cmd: string]: number}): Promise<Section[]> {
+        // Skip if the file has already been parsed. This is to avoid indefinite
+        // loop under the case that A imports B and B imports back A.
+        if (filesBuilt.has(file)) {
+            return []
+        }
+        filesBuilt.add(file)
+
+        // `getDirtyContent` is used here. I did not check if this is
+        // appropriate.
+        const content = this.extension.manager.getDirtyContent(file)
+        if (!content) {
+            this.extension.logger.addLogMessage(`Error loading LaTeX during structuring: ${file}.`)
+            return []
+        }
+
+        // Use `latex-utensils` to generate the AST.
+        const ast = await this.extension.pegParser.parseLatex(content).catch((e) => {
+            if (bibtexParser.isSyntaxError(e)) {
+                const line = e.location.start.line
+                this.extension.logger.addLogMessage(`Error parsing LaTeX during structuring: line ${line} in ${file}.`)
+            }
+            return
+        })
+        if (!ast) {
+            return []
+        }
+
+        // Parse each base-level node. If the node has contents, that function
+        // will be called recursively.
+        let sections: Section[] = []
+        for (const node of ast.content) {
+            sections = [
+                ...sections,
+                ...await this.parseLaTeXNode(node, file, subFile, filesBuilt, commands, depths)
+            ]
+        }
+
+        return sections
+    }
+
+    /**
+     * This function parses a particular LaTeX AST node and its sub-nodes
+     * (contents by `latex-utensils`).
+     *
+     * @param node The AST node to be parsed.
+     *
+     * All other parameters are identical to {@link buildLaTeXSectionFromFile}.
+     *
+     * @returns A flat array of {@link Section} of this node.
+     */
+    private async parseLaTeXNode(node: latexParser.Node, file: string, subFile: boolean, filesBuilt: Set<string>, commands: {cmds: string[], envs: string[], secs: string[]}, depths: {[cmd: string]: number}): Promise<Section[]> {
+        let sections: Section[] = []
+        if (latexParser.isCommand(node)) {
+            if (commands.secs.includes(node.name.replace(/\*$/, ''))) {
+                // \section{Title}
+                sections.push(new Section(
+                    node.name.endsWith('*') ? SectionKind.NoNumberSection : SectionKind.Section,
+                    this.captionify(node.args[node.args.length - 1]),
```

Comment:
I think it would be better to explcitly check that we are in one of the two cases
- `node.args` has length 1 and `node.args[0]` is a `Group`
- `node.args` has length 2 and `node.args[0]` is a `OpptionalArg` and `node.args[1]` is a `Group`